In [28]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [34]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [3]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [4]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [38]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
scaler = StandardScaler()

In [39]:
# Create a DataFrame with the scaled data
# Copy the crypto names from the original data
# Set the coinid column as index
df_market_data_scaled = pd.DataFrame(scaler.fit_transform(df_market_data), columns=df_market_data.columns, index=df_market_data.index)

In [40]:
# Displaying First Five rows of scaled DataFrame
df_market_data_scaled.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Data.

In [41]:
# Create a list with the number of k-values from 1 to 11
klist = [1,2,3,4,5,6,7,8,9,10,11]

In [9]:
# Create an empty list to store the inertia values
inertia_values = []

# Create a for loop to compute the inertia with each possible value of k
for k in klist:  # Adjust the range as needed
    # 1. Create a KMeans model using the loop counter for the n_clusters
    kmeans = KMeans(n_clusters=k, random_state=0)
    
    # 2. Fit the model to the data using `df_market_data_scaled`
    kmeans.fit(df_market_data_scaled)  
    
    # 3. Append the model.inertia_ to the inertia list
    inertia_values.append(kmeans.inertia_)

/Users/kathia/anaconda3/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
/Users/kathia/anaconda3/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
/Users/kathia/anaconda3/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
/Users/kathia/anaconda3/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
/Users/kathia/an

In [43]:
len(inertia_values)

11

In [42]:
inertia_values

[287.0,
 195.82021818036046,
 123.19048183836958,
 79.02243535120975,
 65.302379141625,
 52.88851821977531,
 43.91469044757746,
 37.517032490103574,
 32.485240832183536,
 28.222899290608925,
 24.545512769147678]

In [12]:
import holoviews as hv

# Create a dictionary with the data for the Elbow curve
elbow_data = {
    'Number of Clusters (k)': range(1, len(inertia_values) + 1),
    'Inertia': inertia_values
}

# Create a DataFrame with the data for the Elbow curve
elbow_df = pd.DataFrame(elbow_data)

# Plot a line chart with all the inertia values computed with 
elbow_curve = elbow_df.hvplot.line(x='Number of Clusters (k)', y='Inertia', title='Elbow Curve')

# the different values of k to visually identify the optimal value for k.
elbow_curve

:Curve   [Number of Clusters (k)]   (Inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** 4

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [46]:
# Initialize the K-Means model using the best value for k
kmeans_model = KMeans(n_clusters=4, random_state=0)

In [47]:
# Fit the K-Means model using the scaled data
kmeans_model.fit(df_market_data_scaled)

/Users/kathia/anaconda3/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


KMeans(n_clusters=4, random_state=0)

In [48]:
# Predict the clusters to group the cryptocurrencies using the scaled data
cluster_labels = kmeans_model.predict(df_market_data_scaled)

# Print the resulting array of cluster values.
cluster_labels

array([0, 0, 3, 3, 0, 0, 0, 0, 0, 3, 3, 3, 3, 0, 3, 0, 3, 3, 0, 3, 3, 0,
       3, 3, 3, 3, 3, 3, 0, 3, 3, 3, 1, 0, 3, 3, 2, 3, 3, 3, 3],
      dtype=int32)

In [49]:
# Create a copy of the DataFrame
df_market_data_copy = df_market_data_scaled.copy()

In [50]:
# Add a new column to the DataFrame with the predicted clusters
df_market_data_copy['Cluster'] = cluster_labels

In [51]:
# Display sample data
df_market_data_copy.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  Cluster  
coin_id                                            
bitcoin                        -0.251637        0  
ethereum                       -0.199352        0  
tether                         -0.282061        3  
ripple                         -0.295546        3  
bitcoin-cash                   -0.270317        0

In [53]:
df_market_data_copy["crypto_name"]=df_market_data_copy.index

In [54]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.

scatter_plot = df_market_data_copy.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    c="Cluster",  # Color the points by cluster labels
    cmap='Category20',  # Choose a colormap for the colors
    hover_cols=["crypto_name"],  # Display cryptocurrency names on hover
    size=40  # Adjust point size as needed
)

# Display the scatter plot
scatter_plot

:Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,Cluster,crypto_name)

---

### Optimize Clusters with Principal Component Analysis.

In [56]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

In [57]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
pca_result = pca.fit_transform(df_market_data_scaled)

df_pca = pd.DataFrame(data=pca_result, columns=['PC1', 'PC2', 'PC3'], index=df_market_data_scaled.index)

# View the first five rows of the DataFrame. 
df_pca.head()

PC1       PC2       PC3
coin_id                                   
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715

In [59]:
# Retrieve the explained variance to determine how much information
# can be attributed to each principal component.
explained_variance = pca.explained_variance_ratio_


# Display the explained variance for each component
for i, variance in enumerate(explained_variance):
    print(f"Principal Component {i + 1}: Explained Variance = {variance:.4f}")

Principal Component 1: Explained Variance = 0.3720
Principal Component 2: Explained Variance = 0.3470
Principal Component 3: Explained Variance = 0.1760


In [24]:
total_explained_variance = sum(explained_variance[:3])
total_explained_variance

0.8950316570309842

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** 0.895031657030984

In [25]:
# Create a new DataFrame with the PCA data.

# Creating a DataFrame with the PCA data

# Copy the crypto names from the original data


# Set the coinid column as index


# Display sample data

df_pca.head()

PC1       PC2       PC3
coin_id                                   
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715

---

### Find the Best Value for k Using the PCA Data

In [60]:
# Create a list with the number of k-values from 1 to 11
klist = [1,2,3,4,5,6,7,8,9,10,11]

In [62]:
# Create an empty list to store the inertia values
inertia_values_pca = []

# Create a for loop to compute the inertia with each possible value of k
for k in klist:  # Adjust the range as needed
    # 1. Create a KMeans model using the loop counter for the n_clusters
    kmeans_pca = KMeans(n_clusters=k, random_state=0)
    
    # 2. Fit the model to the PCA-transformed data using `df_market_data_pca`
    kmeans_pca.fit(df_pca)
    
    # 3. Append the model.inertia_ to the inertia list
    inertia_values_pca.append(kmeans_pca.inertia_)

/Users/kathia/anaconda3/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
/Users/kathia/anaconda3/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
/Users/kathia/anaconda3/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
/Users/kathia/anaconda3/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
/Users/kathia/an

In [63]:
inertia_values_pca

[256.8740855678926,
 165.9019940203601,
 93.77462568057297,
 49.66549665179738,
 37.87874703346251,
 27.618971787957445,
 21.182775862957335,
 17.3898232047689,
 13.593650379876738,
 10.55935756279343,
 8.007418066730596]

In [64]:
# Create a dictionary with the data for the Elbow curve
elbow_data_pca = {
    'Number of Clusters (k)': range(1, len(inertia_values_pca) + 1),
    'Inertia': inertia_values_pca
}

# Create a DataFrame with the data for the Elbow curve
elbow_df_pca = pd.DataFrame(elbow_data_pca)

In [65]:
# Plot a line chart with all the inertia values computed with 
# Plot the Elbow curve using hvplot
elbow_curve_pca = elbow_df_pca.hvplot.line(x='Number of Clusters (k)', y='Inertia', title='Elbow Curve (PCA Data)')
# the different values of k to visually identify the optimal value for k.
# Display the interactive Elbow curve plot
elbow_curve_pca

:Curve   [Number of Clusters (k)]   (Inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** 4


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** No

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [66]:
# Initialize the K-Means model using the best value for k
best_kmeans_model = KMeans(n_clusters=4, random_state=0)

In [68]:
# Fit the K-Means model using the PCA data
best_kmeans_model.fit(df_pca)

/Users/kathia/anaconda3/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


KMeans(n_clusters=4, random_state=0)

In [69]:
# Predict the clusters to group the cryptocurrencies using the PCA data
cluster_labels_pca = best_kmeans_model.predict(df_pca)

# Print the resulting array of cluster values.
print(cluster_labels_pca)

[0 0 3 3 0 0 0 0 0 3 3 3 3 0 3 0 3 3 0 3 3 0 3 3 3 3 3 3 0 3 3 3 1 0 3 3 2
 3 3 3 3]


In [71]:
# Create a copy of the DataFrame with the PCA data
df_pca_copy = df_pca.copy()


# Add a new column to the DataFrame with the predicted clusters
df_pca_copy['Predicted_Cluster'] = cluster_labels_pca


# Display sample data
df_pca_copy.head()

PC1       PC2       PC3  Predicted_Cluster
coin_id                                                      
bitcoin      -0.600667  0.842760  0.461595                  0
ethereum     -0.458261  0.458466  0.952877                  0
tether       -0.433070 -0.168126 -0.641752                  3
ripple       -0.471835 -0.222660 -0.479053                  3
bitcoin-cash -1.157800  2.041209  1.859715                  0

In [74]:
# Create a scatter plot using hvPlot by setting 
# `x="PC1"` and `y="PC2"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.


scatter_plot_pca = df_pca_copy.hvplot.scatter(
    x="PC1",
    y="PC2",
    c="Predicted_Cluster",  # Color the points by predicted clusters
    cmap='Category20',  # Choose a colormap for the colors
    hover_cols=["crypto_name"],  # Display cryptocurrency names on hover
    size=40  # Adjust point size as needed
)

# Display the scatter plot
scatter_plot_pca

:Scatter   [PC1]   (PC2,Predicted_Cluster)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [75]:
# Composite plot to contrast the Elbow 
elbow_curve_original = elbow_df.hvplot.line(x='Number of Clusters (k)', y='Inertia', title='Elbow Curve (Original Data)')
elbow_curve_pca = elbow_df_pca.hvplot.line(x='Number of Clusters (k)', y='Inertia', title='Elbow Curve (PCA Data)')

In [76]:
 df_pca_copy["crypto_name"]= df_pca_copy.index

In [77]:
# Composite plot to contrast the clusters
# YOUR CODE HERE!

cluster_plot_original = df_market_data_copy.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    c="Cluster",
    cmap='Category20',
    hover_cols=["crypto_name"],
    title='Cluster Plot (Original Data)'
)
cluster_plot_pca = df_pca_copy.hvplot.scatter(
    x="PC1",
    y="PC2",
    c="Predicted_Cluster",
    cmap='Category20',
    hover_cols=["crypto_name"],
    title='Cluster Plot (PCA Data)'
)

# Create a layout to display both the Elbow curves and cluster plots side by side
layout = (elbow_curve_original + elbow_curve_pca + cluster_plot_original + cluster_plot_pca).cols(2)

# Display the composite plot
layout

:Layout
   .Curve.I    :Curve   [Number of Clusters (k)]   (Inertia)
   .Curve.II   :Curve   [Number of Clusters (k)]   (Inertia)
   .Scatter.I  :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,Cluster,crypto_name)
   .Scatter.II :Scatter   [PC1]   (PC2,Predicted_Cluster,crypto_name)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** Improved Clustering,Dimension Reduction and Loss of Information